Mathlib/AlgebraicGeometry/EllipticCurve/NormalForms.lean

```diff
@@ -0,0 +1,658 @@
+
+namespace WeierstrassCurve
+
+/-! ### Normal forms of characteristic ≠ 2 -/
+
+/-- A `WeierstrassCurve` is in normal form of characteristic ≠ 2, if its $a_1, a_3 = 0$.
+In other words it is $Y^2 = X^3 + a_2X^2 + a_4X + a_6$. -/
+@[mk_iff]
+structure IsCharNeTwoNF : Prop where
+  a₁ : W.a₁ = 0
+  a₃ : W.a₃ = 0
+
+namespace IsCharNeTwoNF
+
+variable {W} (self : W.IsCharNeTwoNF)
+include self
+
+theorem b₂ : W.b₂ = 4 * W.a₂ := by
+  rw [WeierstrassCurve.b₂, self.a₁]
+  ring1
+
+theorem b₄ : W.b₄ = 2 * W.a₄ := by
+  rw [WeierstrassCurve.b₄, self.a₃]
+  ring1
+
+theorem b₆ : W.b₆ = 4 * W.a₆ := by
+  rw [WeierstrassCurve.b₆, self.a₃]
+  ring1
+
+theorem b₈ : W.b₈ = 4 * W.a₂ * W.a₆ - W.a₄ ^ 2 := by
+  rw [WeierstrassCurve.b₈, self.a₁, self.a₃]
+  ring1
+
+theorem c₄ : W.c₄ = 16 * W.a₂ ^ 2 - 48 * W.a₄ := by
+  rw [WeierstrassCurve.c₄, self.b₂, self.b₄]
+  ring1
+
+theorem c₆ : W.c₆ = -64 * W.a₂ ^ 3 + 288 * W.a₂ * W.a₄ - 864 * W.a₆ := by
+  rw [WeierstrassCurve.c₆, self.b₂, self.b₄, self.b₆]
+  ring1
+
+theorem Δ : W.Δ = -64 * W.a₂ ^ 3 * W.a₆ + 16 * W.a₂ ^ 2 * W.a₄ ^ 2 - 64 * W.a₄ ^ 3
+    - 432 * W.a₆ ^ 2 + 288 * W.a₂ * W.a₄ * W.a₆ := by
+  rw [WeierstrassCurve.Δ, self.b₂, self.b₄, self.b₆, self.b₈]
+  ring1
+
+end IsCharNeTwoNF
+
+section VariableChange
+
+variable [Invertible (2 : R)]
+
+/-- There is an explicit change of variables of a `WeierstrassCurve` to
+a normal form of characteristic ≠ 2, provided that 2 is invertible in the ring. -/
+@[simps]
+def toCharNeTwoNF : VariableChange R where
+  u := 1
+  r := 0
+  s := ⅟2 * -W.a₁
+  t := ⅟2 * -W.a₃
+
+theorem toCharNeTwoNF_spec : (W.variableChange W.toCharNeTwoNF).IsCharNeTwoNF := by
+  constructor <;> simp
+
+theorem exists_variableChange_isCharNeTwoNF :
+    ∃ C : VariableChange R, (W.variableChange C).IsCharNeTwoNF :=
+  ⟨_, W.toCharNeTwoNF_spec⟩
+
+end VariableChange
+
+/-! ### Short normal form -/
+
+/-- A `WeierstrassCurve` is in short normal form, if its $a_1, a_2, a_3 = 0$.
+In other words it is $Y^2 = X^3 + a_4X + a_6$.
+
+This is the normal form of characteristic ≠ 2 or 3,
+also the normal form of characteristic = 3 and j = 0. -/
+@[mk_iff]
+structure IsShortNF : Prop where
+  a₁ : W.a₁ = 0
+  a₂ : W.a₂ = 0
+  a₃ : W.a₃ = 0
+
+namespace IsShortNF
+
+variable {W} (self : W.IsShortNF)
+include self
+
+theorem isCharNeTwoNF : W.IsCharNeTwoNF := ⟨self.a₁, self.a₃⟩
+
+theorem b₂ : W.b₂ = 0 := by
+  simp [self.isCharNeTwoNF.b₂, self.a₂]
+
+theorem b₄ : W.b₄ = 2 * W.a₄ := self.isCharNeTwoNF.b₄
+
+theorem b₆ : W.b₆ = 4 * W.a₆ := self.isCharNeTwoNF.b₆
+
+theorem b₈ : W.b₈ = -W.a₄ ^ 2 := by
+  simp [self.isCharNeTwoNF.b₈, self.a₂]
+
+theorem c₄ : W.c₄ = -48 * W.a₄ := by
+  simp [self.isCharNeTwoNF.c₄, self.a₂]
+
+theorem c₆ : W.c₆ = -864 * W.a₆ := by
+  simp [self.isCharNeTwoNF.c₆, self.a₂]
+
+theorem Δ : W.Δ = -16 * (4 * W.a₄ ^ 3 + 27 * W.a₆ ^ 2) := by
+  rw [self.isCharNeTwoNF.Δ, self.a₂]
+  ring1
+
+theorem b₄_of_char_three [CharP R 3] : W.b₄ = -W.a₄ := by
```

Comment:
Extract `[CharP R 3]` into a variable outside since it's used multiple times here?

Same with those in `IsCharThreeJNeZero` (both `W` and `E`) and `IsCharTwoJNeZero`.